File: core/src/engine/general/view/gui/Button.scala
package engine.general.view.gui

import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType
import com.badlogic.gdx.graphics.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.geom.Rectangle2D
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.graphics.g2d.BitmapFont


object Button{
	class ButtonBuilder(){
		var x:Int=0
		var y:Int=0
		var text:String=""
		var width:Int=90
		var height:Int=25
		var font:BitmapFont=null
		
		def x(xT:Int):ButtonBuilder={
			x=xT
			return this
		}
		
		def y(yT:Int):ButtonBuilder={
			y=yT
			return this
		}
		
		def text(t:String):ButtonBuilder={
			text=t
			return this
		}
		
		def width(w:Int):ButtonBuilder={
			width=w
			return this
		}
		
		def height(h:Int):ButtonBuilder={
			height=h
			return this
		}
		
		def font(f:BitmapFont):ButtonBuilder={
		    font=f
		    return this
		}
		
		def build():Button={
		    return new Button(this)
		}
	}
}
class Button(x:Integer,y:Integer,builder:Button.ButtonBuilder) extends GuiItem(x,y){

	val width=builder.width
	val height=builder.height
	var text=builder.text 
	val font=builder.font 
	
    val myColor=Color.WHITE
    def getWidth=width
    
    var drawRect=new Rectangle2D.Double(xPos.toDouble,(yPos-height/2).toDouble,width.toDouble,height.toDouble)

	def this(builder:Button.ButtonBuilder){
	    this(builder.x,builder.y,builder)
	}

    def draw(render:ShapeRenderer,batch:SpriteBatch){
      
    	render.begin(ShapeType.Filled)
    	render.setColor(Color.GRAY)
    	render.rect(xPos,yPos,drawRect.width.toFloat,drawRect.height.toFloat)
    	render.end()
    	
    	render.begin(ShapeType.Line)
    	render.setColor(0,0,0,1)
    	render.rect(xPos,yPos,drawRect.width.toFloat,drawRect.height.toFloat)
    	render.end()
    	
    	batch.begin()
    	font.setColor(Color.WHITE)
    	font.draw(batch,text,xPos,yPos+height)
    	batch.end()
    }

    def getText=text
    
    def setText(s:String){
        text=s
    }
    
    def contains(x:Int,y:Int):Boolean=drawRect.contains(x.toDouble,y.toDouble)
}

File: core/src/engine/general/view/gui/Button.java
package engine.general.view.gui;

import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType;
import com.badlogic.gdx.graphics.Color;
import java.awt.geom.Rectangle2D;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.graphics.g2d.BitmapFont;


public class Button extends GuiItem{
	
	public static class ButtonBuilder{
	  
		int x=0;
		int y=0;
		String text="";
		int width=90;
		int height=25;
		BitmapFont font=null;
		
		public ButtonBuilder x(int xT){
			x=xT;
			return this;
		}
		
		public ButtonBuilder y(int yT){
			y=yT;
			return this;
		}
		
		public ButtonBuilder text(String t){
			text=t;
			return this;
		}
		
		public ButtonBuilder width(int w){
			width=w;
			return this;
		}
		
		public ButtonBuilder height(int h){
			height=h;
			return this;
		}
		
		public ButtonBuilder font(BitmapFont f){
		    font=f;
		    return this;
		}
		
		public Button build(){
		    return new Button(this);
		}
	}
	
	final int width;
	final int height;
	private String text;
	final BitmapFont font;
	final Color myColor=Color.WHITE;
	
	private Rectangle2D.Double drawRect;
	
	private Button(ButtonBuilder b){
		super(b.x,b.y);
		width=b.width;
		height=b.height;
		text=b.text;
		font=b.font;
	    drawRect=new Rectangle2D.Double((double)xPos,(double)(yPos-height/2),(double)width,(double)height);
	}
	
	public int getWidth(){
		return width;
	}
    
    public void draw(ShapeRenderer render,SpriteBatch batch){
      
    	render.begin(ShapeType.Filled);
    	render.setColor(Color.GRAY);
    	render.rect(xPos,yPos,(float)drawRect.width,(float)drawRect.height);
    	render.end();
    	
    	render.begin(ShapeType.Line);
    	render.setColor(0,0,0,1);
    	render.rect(xPos,yPos,(float)drawRect.width,(float)drawRect.height);
    	render.end();
    	
    	batch.begin();
    	font.setColor(Color.WHITE);
    	font.draw(batch,text,xPos,yPos+height);
    	batch.end();
    }

    public String getText(){
    	return text;
    }
    
    public void setText(String s){
        text=s;
    }
    
    public boolean contains(int x,int y){
    	return drawRect.contains((double)x,(double)y);
    }
}